docs/ConstPropagationPass.md

```diff
@@ -0,0 +1,266 @@
+
+## Example
+Given the following code:
+```mlir
+func @foo() -> tensor<1xf32> {
+  %0 = "onnx.Constant"() {value = dense<[1.0]> : tensor<1xf32>} : () -> tensor<1xf32>
+  %1 = "onnx.Constant"() {value = dense<[2.0]> : tensor<1xf32>} : () -> tensor<1xf32>
+  %2 = "onnx.Add"(%0, %1) : (tensor<1xf32> , tensor<1xf32>) -> tensor<1xf32>
+  %3 = "onnx.Constant"() {value = dense<[3.0]> : tensor<1xf32>} : () -> tensor<1xf32>
+  %4 = "onnx.Add"(%2, %3) : (tensor<1xf32> , tensor<1xf32>) -> tensor<1xf32>
+  "std.return"(%4) : (tensor<1xf32>) -> ()
+}
+```
+
+If we call `onnx-mlir-op --constprop-onnx`, we will get:
+```mlir
+func @foo() -> tensor<1xf32> {
+  %0 = "onnx.Constant"() {value = dense<[6.0]> : tensor<1xf32>} : () -> tensor<1xf32>
+  "std.return"(%0) : (tensor<1xf32>) -> ()
+}
+```
+
+## Remark 
+
+ONNXConstantOp uses MLIR DenseElementsAttr to store constant values. It is
+important to note that, once a DenseElementsAttr is created, it is alive and 
+consumes memory until the end of compilation. In [Example](#example), all the
+three DenseElementsAttrs in the three ONNXConstantOps exist until the end of
+compilation. Especially, two intermediate DenseElementsAttrs in the two
+ONNXConstantOps produced by folding the two ONNXAddOps also exist. For a
+practice model, the number of intermediate DenseElementsAttrs will increase
+quickly, which lead to a large memory footprint during compilation. 
+
+To avoid creating too many DenseElementsAttrs for intermediate ONNXConstant ops
+during `--constprop-onnx`, we design a mechanism that dynamically allocates and
+deallocates buffers for intermediate ONNXConstantOps and only creates
+DenseElementsAttr for the final results of constant propagation.
+
+In particular, we maintain a buffer pool internally. When an ONNXConstantOp
+is reached at the first time, we read its DenseElementsAttr and store
+data to an array buffer in the pool. A unique buffer ID is used to map the
+ONNXConstantOp to the buffer. All constant computations are then done on array
+buffers, which is to avoid creating DenseElementsAttr for intermediate
+ONNXConstantOps. Buffers are automatically freed if they are not used.
+
+We provide three helper functions to use when working with buffers:
+1. `getArrayFromAttributeOrBuffer(PatternRewriter &rewriter, Operation *op)`
+   - create a buffer from a dense attribute at the first time we reach the
+     const 'op' and add the buffer to the buffer pool, or
+   - get the buffer from the buffer pool if it was created.
+2. `createConstantOpAndStoreBufferPtr(..., char *buffer)`
+   - create a new ONNXConstantOp using the given buffer, and
+   - add the buffer to the buffer pool.
+3. `allocateBufferFor(Value value, bool useMaxSize = false)`
+   - create a new buffer whose size is obtained from the type of 'value'.
+
+Note that:
+  - The buffers in the buffer pool will be automatically freed. Users don't
+    need to take care about that.
```

Comment:
Can you add a quick explanation of how the freeing is occurring? It is freed if its associated ONNXConstantOp is deleted?

Reply:
> It is freed if its associated ONNXConstantOp is deleted?

Yes. I added an explanation.

Reply:
tx